app/services/cache_manager.rb

```diff
@@ -10,7 +10,39 @@ def self.create_new_permission_cache
   end
 
   def self.create_new_filter_cache
-    Filter.all.each(&:save!)
+    locked_roles = Role.select(&:locked?)
+    origins = locked_roles.map(&:origin)
+    locked_roles.each { |role| role.origin = nil }
+    locked_roles.each { |role| role.modify_locked = true }
+    locked_roles.each(&:save)
+    Filter.all.each do |filter|
+      next if filter.save
+
+      if filter.errors[:permissions] == ['Permissions must be of same resource type']
+        # If the filter was created long time ago, it may have had several
+        # resources of the same type. To fix it, let's split it into different
+        # filters each with its own resource type.
+        filter.permissions.group_by(&:resource_type).each do |resource_type, permissions|
+          # Create a new filter for all resource types
+          restricted_resource_type_filter = Filter.new(
+            :role => filter.role,
+            :permissions => permissions,
+            :taxonomy_search => filter.taxonomy_search,
+            :organizations => filter.organizations,
+            :locations => filter.locations,
+            :override => filter.override
+          )
+          restricted_resource_type_filter.save
+        end
+
+        # At this point, all permissions are in their own filters by resource type
+        filter.errors.clear
+        filter.destroy
+      end
+    end
+    locked_roles.each_with_index { |role, index| role.origin = origins[index] }
+    locked_roles.each { |role| role.modify_locked = false }
+    locked_roles.each(&:save)
```

Comment:
I'm not really sure, but I think calling `.save` is what we want - so that the related objects (Filterings, Permissions, Role, UserRole..) are also running callbacks. At least for `create_new_permission_cache` the point is precisely to run callbacks on all the objects related with the saved UsergroupMember and UserRole. I think `.save` is also future-proof. 

Disabling the lock validation while in Rake isn't ideal, as tests would have to handle that somehow too, not to mention seeds could also accidentally update locked roles, etc...